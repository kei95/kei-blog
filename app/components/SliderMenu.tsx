import { useState } from "react";
import RouteLink from "~/lib/RouteLink";

export const SLIDER_MENU_TEST_ID = "SLIDER_MENU_TEST_ID";
type SliderState = "OPEN" | "CLOSED";

export default function SliderMenu() {
  const [sliderMenuState, setSliderMenuState] = useState<SliderState>("CLOSED");

  const sliderWidth = sliderMenuState === "OPEN" ? "w-56" : "w-0";

  const openSliderMenu = () => {
    setSliderMenuState("OPEN");
  };

  const closeSliderMenu = () => {
    setSliderMenuState("CLOSED");
  };

  const handleOnclickLink = () => {
    closeSliderMenu();
  };

  return (
    <div data-testid={SLIDER_MENU_TEST_ID}>
      <button className="md:hidden" type="button" onClick={openSliderMenu}>
        <svg
          xmlns="http://www.w3.org/2000/svg"
          className="h-6 w-6"
          fill="none"
          viewBox="0 0 24 24"
          stroke="currentColor"
          strokeWidth="2"
        >
          <path
            strokeLinecap="round"
            strokeLinejoin="round"
            d="M4 6h16M4 12h16m-7 6h7"
          />
        </svg>
      </button>

      {/* Slider menu - only shows up in narrow screens */}
      <aside
        className={`overflow-x-hidden right-0 duration-300 fixed top-0 h-screen bg-background_darken ${sliderWidth}`}
      >
        <div className="flex flex-col">
          <button type="button" onClick={closeSliderMenu}>
            &times;
          </button>
          <RouteLink onClick={handleOnclickLink} to="/">
            Home
          </RouteLink>
          <RouteLink onClick={handleOnclickLink} to="/#about">
            About
          </RouteLink>
          <RouteLink onClick={handleOnclickLink} to="/#career">
            Career
          </RouteLink>
          <RouteLink onClick={handleOnclickLink} to="/#posts">
            Posts
          </RouteLink>
        </div>
      </aside>
    </div>
  );
}
