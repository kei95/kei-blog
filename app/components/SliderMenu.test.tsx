import { render, screen } from "@testing-library/react";

import SliderMenu, { SLIDER_MENU_TEST_ID } from "./SliderMenu";

describe("<SliderMenu />", () => {
  Object.defineProperties(window.HTMLElement.prototype, {
    offsetWidth: {
      get: function () {
        return "500px";
      },
    },
  });

  test("render the component - it should have expected elements", () => {
    render(<SliderMenu />);
    expect(screen.getByTestId(SLIDER_MENU_TEST_ID)).toMatchSnapshot();
  });

  test("check each link buttons - they should be tagged to expected section", () => {
    render(<SliderMenu />);
    const expectedLinks = [
      { to: "/", name: "Home" },
      { to: "/#about", name: "About" },
      { to: "/#career", name: "Career" },
      { to: "/#posts", name: "Posts" },
    ];

    // Iterate over each links - it should find all expected links with its direction value
    for (let i = 0; i < expectedLinks.length; i++) {
      const { to, name } = expectedLinks[i];
      const linkEl = screen.getByRole("link", { name });

      expect(linkEl).toHaveAttribute("href", `${to}`);
    }
  });
});
