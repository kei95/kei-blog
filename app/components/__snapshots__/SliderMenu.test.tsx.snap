// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SliderMenu /> render the component - it should have expected elements 1`] = `
<div
  class="md:hidden"
  data-testid="SLIDER_MENU_TEST_ID"
>
  <button
    type="button"
  >
    <svg
      class="h-6 w-6"
      fill="none"
      stroke="currentColor"
      stroke-width="2"
      viewBox="0 0 24 24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4 6h16M4 12h16m-7 6h7"
        stroke-linecap="round"
        stroke-linejoin="round"
      />
    </svg>
  </button>
  <aside
    class="overflow-x-hidden right-0 duration-300 fixed top-0 h-screen bg-background_darken w-0"
  >
    <ul
      class="inline-block"
    >
      <li>
        <button
          class="px-4 p-1 text-4xl hover:opacity-60"
          type="button"
        >
          Ã—
        </button>
      </li>
      <li>
        <a
          href="/"
        >
          <h3
            class="bg-clip-text text-transparent bg-white hover:bg-gradient-to-r from-[#edca85] via-[#d55594] to-[#5594d5] px-4 py-1"
          >
            Home
          </h3>
        </a>
      </li>
      <li>
        <a
          href="/#about"
        >
          <h3
            class="bg-clip-text text-transparent bg-white hover:bg-gradient-to-r from-[#edca85] via-[#d55594] to-[#5594d5] px-4 py-1"
          >
            About
          </h3>
        </a>
      </li>
      <li>
        <a
          href="/#career"
        >
          <h3
            class="bg-clip-text text-transparent bg-white hover:bg-gradient-to-r from-[#edca85] via-[#d55594] to-[#5594d5] px-4 py-1"
          >
            Career
          </h3>
        </a>
      </li>
      <li>
        <a
          href="/#posts"
        >
          <h3
            class="bg-clip-text text-transparent bg-white hover:bg-gradient-to-r from-[#edca85] via-[#d55594] to-[#5594d5] px-4 py-1"
          >
            Posts
          </h3>
        </a>
      </li>
    </ul>
  </aside>
</div>
`;
